File: src/mod_ident.rs
use crate::Version;
use anyhow::anyhow;
use serde::Deserialize;
use std::fmt;
use std::str::FromStr;

#[derive(Clone, Eq, Debug, Deserialize, PartialEq)]
pub struct ModIdent {
    pub name: String,
    pub version: Option<Version>,
}

impl ModIdent {
    /// In cases where `version` is guaranteed to be Some(), get the version
    pub fn get_guaranteed_version(&self) -> &Version {
        self.version
            .as_ref()
            .expect("Version was not present in guaranteed case")
    }
}

impl FromStr for ModIdent {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts: Vec<&str> = s.split('@').collect();
        match parts[..] {
            [name] => Ok(Self {
                name: name.to_string(),
                version: None,
            }),
            [name, version] => {
                let parsed_version = Version::from_str(version);
                if let Ok(version) = parsed_version {
                    // Validate that the version does *not* have prerelease or build data
                    Ok(Self {
                        name: name.to_string(),
                        version: Some(version),
                    })
                } else {
                    Err(anyhow!(
                        "Invalid version identifier: {}",
                        version.to_string()
                    ))
                }
            }
            _ => Err(anyhow!(
                "Incorrect mod format - must be 'Name' or 'Name@Version'"
            )),
        }
    }
}

impl fmt::Display for ModIdent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{}",
            self.name,
            match &self.version {
                Some(version_req) => format!(" {}", version_req),
                _ => "".to_string(),
            }
        )
    }
}

File: lib/modident.go
package fmm

import (
	"strings"
)

// A small representation of a mod, with an optional version.
type ModIdent struct {
	Name    string
	Version *Version
}

// Returns a ModIdent parsed from an input string with the format of 'name',
// 'name_version', or 'name_version.zip'.
func NewModIdent(input string) ModIdent {
	input = strings.TrimSuffix(input, ".zip")
	parts := strings.Split(input, "_")
	if len(parts) == 1 {
		return ModIdent{input, nil}
	}

	name := strings.Join(parts[:len(parts)-1], "_")
	version, err := NewVersion(parts[len(parts)-1])
	if err != nil {
		return ModIdent{input, nil}
	}
	return ModIdent{name, version}
}

// Returns a string in the format of 'name' or 'name_version'.
func (i *ModIdent) ToString() string {
	if i.Version != nil {
		return i.Name + " " + i.Version.ToString(false)
	}
	return i.Name
}
